lib/datasource/aws-machine-image/index.ts

```diff
@@ -0,0 +1,114 @@
+import { DescribeImagesCommand, EC2Client, Image } from '@aws-sdk/client-ec2';
+import { cache } from '../../util/cache/package/decorator';
+import * as amazonMachineImageVersioning from '../../versioning/aws-machine-image';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, ReleaseResult } from '../types';
+
+export class AwsMachineImageDataSource extends Datasource {
+  static readonly id = 'aws-machine-image';
+
+  override readonly defaultVersioning = amazonMachineImageVersioning.id;
+
+  override readonly caching = true;
+
+  override readonly defaultConfig = {
+    // Because AMIs don't follow any versioning scheme, we override commitMessageExtra to remove the 'v'
+    commitMessageExtra: 'to {{{newVersion}}}',
+    prBodyColumns: ['Change', 'Image'],
+    prBodyDefinitions: {
+      Image: '```{{{newDigest}}}```',
+    },
+    digest: {
+      // Because newDigestShort will allways be 'amazon-' we override to print the name of the AMI
+      commitMessageExtra: 'to {{{newDigest}}}',
+      prBodyColumns: ['Image'],
+      prBodyDefinitions: {
+        Image: '```{{{newDigest}}}```',
+      },
+    },
+  };
+
+  readonly ec2: EC2Client;
+
+  readonly now: number;
+
+  constructor() {
+    super(AwsMachineImageDataSource.id);
+    this.ec2 = new EC2Client({});
+    this.now = Date.now();
+  }
+
+  @cache({
+    namespace: `datasource-${AwsMachineImageDataSource.id}`,
+    key: (serializedAmiFilter: string) =>
+      `getSortedAwsMachineImages:${serializedAmiFilter}`,
+  })
+  async getSortedAwsMachineImages(
```

Comment:
Do we really need to have this public? can't we simply test it via the normal public api?

Reply:
☝️ ❓

Reply:
Hi @markussiebert Can you respond to the questions raised by the maintainer/reviewer? I'll add the link so you can find it easily, it's easy to miss in the sea of review comments:

https://github.com/renovatebot/renovate/pull/11869#discussion_r750115854

Reply:
omg, sorry ... I answered, but had to submit my answers as a review? This differs so much from what I am used to with gitlab ...

Reply:
I think the cache decorator should also work on private methods as there is no difference to public at runtime.

So this is only required by direct testing, which is ok for now, as i makes tests easier.

BTW: Caching is disabled on tests and don't need to be tested. We have some special tests to verify caching 😉

Reply:
@viceice  I tried out locally to make this method private. Codecoverage was still 100%, but the decorator wasn't working anymore.

Reply:
@JamieMagee do you why it's not working for private methods?

Reply:
You're correct that private modifiers don't exist at runtime, but the typescript compiler still enforces them.

For example, I've got the `jenkins-plugin` datasource open right now. If I make one of the decorated methods private, I get this error: `Argument of type '"getJenkinsPluginInfo"' is not assignable to parameter of type 'keyof JenkinsPluginsDatasource'.ts(2345)`

That's because of this code:

https://github.com/renovatebot/renovate/blob/789caadbb49e232ef74f59f5c41c1a6fbffee5c0/lib/util/cache/package/decorator.ts#L6-L10

The decorator can decorate methods that are in `keyof U`, in this case `keyof JenkinsPluginsDatasource`.

I can't think of any way that would allow us to decorate private methods, but still retain type safety.

Reply:
There is an open discussion in the TypeScript repo about this scenario https://github.com/microsoft/TypeScript/issues/22677